packages/url_launcher/url_launcher_ios/ios/Classes/URLLauncherPlugin.swift

```diff
@@ -0,0 +1,115 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import Flutter
+
+public final class URLLauncherPlugin: NSObject, FlutterPlugin, UrlLauncherApi {
+
+  public static func register(with registrar: FlutterPluginRegistrar) {
+    let plugin = URLLauncherPlugin()
+    UrlLauncherApiSetup.setUp(binaryMessenger: registrar.messenger(), api: plugin)
+    registrar.publish(plugin)
+  }
+
+  private var currentSession: URLLaunchSession?
+  private let launcher: Launcher
+
+  init(launcher: Launcher = UIApplicationLauncher()) {
+    self.launcher = launcher
+  }
+
+  func canLaunchUrl(url: String) -> Bool {
+    guard let url = URL(string: url) else { return false }
+    return launcher.canOpenURL(url)
+  }
+
+  func launchUrl(
+    url: String, universalLinksOnly: Bool, completion: @escaping (Result<Bool, Error>) -> Void
+  ) {
+    guard let url = URL(string: url) else {
+      completion(Result.failure(invalidURLError(for: url)))
+      return
+    }
+    let options = [UIApplication.OpenExternalURLOptionsKey.universalLinksOnly: universalLinksOnly]
+    launcher.openURL(url, options: options) { success in
+      completion(Result.success(success))
+    }
+  }
+
+  func openUrlInSafariViewController(
+    url: String, completion: @escaping (Result<Bool, Error>) -> Void
+  ) {
+    guard let url = URL(string: url) else {
+      completion(Result.failure(invalidURLError(for: url)))
+      return
+    }
+
+    currentSession = URLLaunchSession(url: url, completion: completion)
+    guard let session = currentSession else { return }
+
+    session.didFinish = { [weak self] in
+      self?.currentSession = nil
+    }
+    topViewController?.present(session.safari, animated: true, completion: nil)
+  }
+
+  func closeSafariViewController() {
+    currentSession?.close()
+  }
+
+  var topViewController: UIViewController? {
+    // TODO(stuartmorgan) Provide a non-deprecated codepath. See
+    // https://github.com/flutter/flutter/issues/104117
+    UIApplication.shared.keyWindow?.rootViewController?.topViewController
+  }
+
+  /**
+    * Creates an error for an invalid URL string.
+    *
+    * @param url The invalid URL string
+    * @return The error to return
+    */
+  func invalidURLError(for url: String) -> Error {
+    GeneralError(
+      code: "argument_error", message: "Unable to parse URL", details: "Provided URL: \(url)")
+  }
+}
+
+/// This method recursively iterate through the view hierarchy
+/// to return the top most view controller.
+///
+/// It supports the following scenarios:
+///
+/// - The view controller is presenting another view.
+/// - The view controller is a UINavigationController.
+/// - The view controller is a UITabBarController.
+///
+/// @return The top most view controller.
+extension UIViewController {
+  var topViewController: UIViewController {
```

Comment:
How does Swift handle name conflicts in extensions? This would be extremely dangerous in Obj-C code; I'm not sure if that's true in Swift as well.

Specifically:
- What would happen if Apple added a `UIViewController.topViewController` in a future update?
- What would happen if another plugin had a `UIViewController.topViewController` extension?

Reply:
In this example, the definition in current module shadows definitions from other modules. So if Apple or another frameworks with the same definitions are both imported, the current module still takes precedence. 

However, if we were to make this extension `public` (rather than `internal`) and use it in other modules, it would be problematic when multiple modules with the same extension are imported, because there would be no way to indicate which one to use. In that case, we can use a prefix to disambiguate.

Reply:
Thanks!